## Alignment to reference genomes

I align the sequencing reads to the newly generated *Lynx rufus* reference genome ([mLynRuf2.2](https://denovo.cnag.cat/lynx_rufus)).

From our previous work ([Kleinmann-Ruiz et al. 2022](https://www.pnas.org/doi/abs/10.1073/pnas.2110614119) and [Bazzicalupo et al. 2023](https://onlinelibrary.wiley.com/doi/full/10.1111/eva.13570)) we have already generated sequences of most samples (see [samples_table](data/samples_table.xlsx)).

Additionally I used sequences generated for the new Plan Nacional project. Libraries were generated by Luc√≠a and Laura and sequenced by NovoGene *check with them for additional details*

### Data Quality Control and Read Processing

I run [fastqc](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) to check the quality of rawreads from sequencing.

Seeing the warnings I receive from fastqc reports I need to process the rawreads to try to remove them.

I use [fastp](https://github.com/OpenGene/fastp) from [Chen et al. (2018)](https://academic.oup.com/bioinformatics/article/34/17/i884/5093234?login=true) with default settings plus the following flags:
```
--dont_overwrite (protect the existing files not to be overwritten by fastp)
--trim_poly_g (detect the polyG in read tails and trim them)
--length_required 30 (reads shorter than length_required will be discarded)
--correction (enable base correction in overlapped regions)
--detect_adapter_for_pe (enable adapter sequence auto-detection)
--thread 6 (worker thread number)
```
 The [sbatch_fastp_fqdir_r1fq_r2fq](src/fastp/sbatch_fastp_fqdir_r1fq_r2fq.sh) script is used together with the [fqdir_r1fq_r2fq_table](config/fastp/fqdir_r1fq_r2fq_table.txt) table to run fastp for all fastq files of all of the samples.
```
for row in $(cat config/fastp/fqdir_r1fq_r2fq_table.txt | tr ' ' ':'); do
    
    fastq_dir=$(echo $row | cut -d':' -f2)
    r1_fastq=$(echo $row | cut -d':' -f3)
    r2_fastq=$(echo $row | cut -d':' -f4)
    echo "sbatch src/fastp/sbatch_fastp_fqdir_r1fq_r2fq.sh ${fastq_dir} ${r1_fastq} ${r2_fastq}"
    sbatch src/fastp/sbatch_fastp_fqdir_r1fq_r2fq.sh ${fastq_dir} ${r1_fastq} ${r2_fastq}

done
```
Some samples were already trimmed with fastp. These are the samples from the new Plan Nacional project trimmed by [Lucia](https://github.com/luciamayorf/Data_preprocessing_alignment?tab=readme-ov-file#2-reads-trimming-and-quality-control) and the samples with higher read depth from previous projects (see [Lynx_Demography](https://github.com/Enricobazzi/Lynx_Demography/blob/main/README.md#run-fastp-on-raw-reads) folder).

### Run alignments

I align the my samples to the [*Lynx rufus* reference genome](https://denovo.cnag.cat/lynx_rufus).

A script will be run for each sample that will use [bwa v0.7.17](https://bio-bwa.sourceforge.net/bwa.shtml), [samtools v1.9](https://www.htslib.org/doc/samtools.html), [picard v2.25.5](https://broadinstitute.github.io/picard/) and [gatk v3.7-0](https://gatk.broadinstitute.org/hc/en-us) that will do the following:
- Align each of the sample's R1-R2 fastq pairs to the reference genome using BWA-MEM and Samtools view
- samtools sort to sort the reads in the bam file
- picardtools AddOrReplaceReadGroups to add read groups to the reads in the bam file
- samtools merge to merge the bam files from the multiple R1-R2 pairs of the sample if necessary
- picardtools MarkDuplicates to mark duplicate reads in the bam
- realign indels with GATK which comprises:
    - gatk RealignerTargetCreator to identify realign targets
    - gatk IndelRealigner to realign the indels
- samtools index to index the indel realigned bam for downstream analyses

To prepare the reference genome of the alignment pipeline I run the following bwa and samtools commands:
```
bwa index mLynRuf2.2.revcomp.scaffolds.fa
samtools faidx mLynRuf2.2.revcomp.scaffolds.fa
samtools dict mLynRuf2.2.revcomp.scaffolds.fa -o mLynRuf2.2.revcomp.scaffolds.dict
```

The script [sbatch_alignment_of_sample_from_yaml](src/alignment/sbatch_alignment_of_sample_from_yaml.sh) will perform these steps for a particular sample,taking the information from a yaml file which is passed as first positional argument. Yaml files of each sample can be found in the [config/alignment](config/alignment/) folder.
```
for yaml in $(ls config/alignment/); do

    echo $yaml
    sbatch src/alignment/sbatch_alignment_of_sample_from_yaml.sh config/alignment/$yaml

done
```